### DISCUSSION TITLE
Proposal for adding Radicle forge

### DISCUSSION THREAD

**Discussion by mzampetakis** [woodpecker-ci/woodpecker#2355] [Request a Feature]
### Clear and concise description of the problem

[Radicle](https://radicle.xyz/) is a sovereign peer-to-peer network for code collaboration, built on top of Git. Radicle enables users to run their own nodes, ensuring censorship-resistant code collaboration and fostering a resilient network without reliance on third-parties. It is under heavy development and currently in alpha version. Apart from the development of the protocol itself it is pretty crucial for radicle community to integrate it with popular and well-established tooling around it so that it will offer developers an easier adaptation process and a complete solution. Having this integration in place will allow even more developers to use woodpecker-ci within their radicle repos. Apart from this, developers that have already setup their CI pipelines in woodpecker will be able to transfer their repos to Radicle and keep their jobs running with minimum effort! With this integration Woodpecker will also become one of the first CI systems to be integrated with Radicle and can position itself as a solution that supports user sovereignty and isn't just compatible with centralized solutions. 

I would very much like to work on this integration / implementation and I was wondering if the maintainers will be open in accepting a PR and perhaps providing some help along the way ? 

### Suggested solution

I will implement a new forge for [Radicle](https://radicle.xyz/) as the [existing ones](https://github.com/woodpecker-ci/woodpecker/tree/main/server/forge) alongside all the official documentation and tutorials required to support it. 
As radicle has a different philosophy compared to the already implemented forges there are some limitations. While some issue might arise during the development of this integration there will be significant support form the radicle team to overcome them when feasible. Here are some of them that I have already identified:
1. Radicle does not support oAuth. Having the nodes properly set up, repos will be available to each connected node, so there is no necessity for authentication except when "enabling" a repo (check #2). Authentication can be achieved through auth keys (public/private).
2. Enabling a repo at woodpecker is the process where the hooks at the specific repo are registered. This action is not (still) available through an API call at the Radicle side. This action is currently supported by adding a specific file within the radicle node so it can be accomplished as a manual step.

### Validations

- [X] Checked that the feature isn't part of the `next` version already [https://woodpecker-ci.org/faq#which-version-of-woodpecker-should-i-use]
- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by anbraten** [comment]
Radicle seems quite interesting. We are open to adding support for it, however if larger adjustments to the fundamentals of Woodpecker will be required we need to check again if its worth the effort and maintenance based on the community interest.

As you mentioned correctly already there are some main interfaces of Woodpecker and the forges:
- oAuth for access
- webhooks to trigger pipelines after someone pushed / tagged / ... code
- the clone plugin to get the repo data to execute pipelines on
- some api calls like listing all branches / PRs, getting a file which could be ignored for the beginning

Some questions I currently don't understand from how radicle works / your proposal:
- radicle uses multiple nodes all being fully indepenent, however woodpecker needs to talk to some node at the end to call the api, clone the repo etc, right? Couldn't there also be an oAuth endpoint on a local node doing local oAuth (like mastodon is doing it for example)? To which api is the radicle UI talking to and how can users authenticate in the UI?
- Webhooks will be triggered by a specific node, but for now you need to configure it by committing sth like a git-hook to a node?
- How would you clone a repo for the pipeline? Looks like there is a special git wrapper is used. So you would need a special clone plugin as well, I guess.

**Reply by mzampetakis** [reply]
It is indeed a very interesting project and that's why I am considering adding this forge to woodpecker! I believe that it's desirable not to make any adjustments to the fundamentals of Woodpecker. But if it is required, I would idealy alter them to the bare minimum. 

As Radicle is a peer-to-peer network for code collaboration it is mandatory to use a node for any kind of operation. For this, I am planning to use a dedicated Radicle node, connected to woodpecker-ci, in order to conduct any task (list repos, clone repo, etc). This node would be something like the "woodpecker node" so that it will be accredible for every task conducted by woodpecker-ci.
So regarding the questions you mentioned:
- All this can be achieved directly through the woodpecker node as I mentioned before.  Radicle UI is directly connected to a node that runs on the same machine.
- This is exactly the scenario. The woodpecker node should be appropriately configured in order to issue any webhook.
- Radicle supports git protocol as we know it. So you can clone a repo either through rad protocol (`rad clone rad:z3gqcJUoA1n9HaHKufZs5FCSGazv5`) either through git (`git clone https://seed.radicle.xyz/z3gqcJUoA1n9HaHKufZs5FCSGazv5.git heartwood`).

**Reply by 6543** [reply]
well if there should be a native rad plugin ... we just have to build one.

the vcs type is stored per repo in database ... so we can even add a "default rad clone" step as we do with git based on that

**Reply by mzampetakis** [reply]
That could be used too @6543 ! We will have to use rad commands instead of the git one by adding support for another SCM in the woodpecker's pipeline code. Having this in place it might be easier in the future to add other SCM systems other that git based ones.

**Reply by gsaslis** [reply]
Hi folks! 

I'm on the Radicle core team and I've discussed with @mzampetakis a little about this integration. 

Please allow me to add some context / direction: 

> radicle uses multiple nodes all being fully indepenent, however woodpecker needs to talk to some node at the end to call the api, clone the repo etc, right? 

That is correct, yes. Each node can (choose to) deploy an API side-by-side with the node that participates in the gossip protocol of the p2p network, so that information stored in that particular node is accessible through that API. 

> Couldn't there also be an oAuth endpoint on a local node doing local oAuth (like mastodon is doing it for example)? 

Perhaps - that isn't something that's available right now though, but considering this is a machine-to-machine integration, it seems to me something like a long-lived access token could be a good starting point. We would need to add support for that on the Radicle side. I created an [issue](https://app.radicle.xyz/nodes/seed.radicle.xyz/rad:z3gqcJUoA1n9HaHKufZs5FCSGazv5/issues/933376a1094158a7085f7c83975fe60ab339bf1d) for that, but also open to other suggestions. 

> To which api is the radicle UI talking to 

see above ;) 

> and how can users authenticate in the UI?

For the radicle web UI in particular, there is a `rad web` command - offered as part of the radicle CLI that handles creating a short-lived (1 week) token that the UI can then use to access the API. 

> Webhooks will be triggered by a specific node, but for now you need to configure it by committing sth like a git-hook to a node?

We'll be building a webhooks mechanism on our side to support this integration - and others too - so we can avoid all this manual hook setup. This should really be supported by the forge, In My Humble Opinion (IMHO), and not be a burden for developers who host their projects there. 

> Radicle supports git protocol as we know it. So you can clone a repo either through rad protocol (rad clone rad:z3gqcJUoA1n9HaHKufZs5FCSGazv5) either through git (git clone https://seed.radicle.xyz/z3gqcJUoA1n9HaHKufZs5FCSGazv5.git heartwood).

I think I would lean towards just using `git` for cloning. 
Any `rad` CLI action requires a radicle identity (key pair) that needs to somehow have been generated and stored / passed to the worker who is running the clone action. I'm sure it will be _possible_ to do that, but I expect it won't be so convenient as to pass an access token and use the HTTP API instead for any radicle-domain-specific actions that are not available with `git` tooling (e.g. updating the build status in the Radicle Patch).

**Reply by mzampetakis** [reply]
We have been discussing about this implementation and the possible issues that we will face. Here is a more complete and up to date list of the design decisions we have come up with:

1. We will use a dedicated Radicle node for the Woodpecker and access it through API.
2. A long lived API token will be used for authenticating Woodpecker to the Radicle node (no oAuth Process will be used).
3. We will have the webhooks mechanism in-place at the Radicle side so Woodpecker will be able to receive any event (push or PR/MR - At Radicle’s side PRs/MRs are called patches) that takes place within a repo.
5. The mechanism for registering any webhook at Redicle’s side will not be available through an API (at least for the first version), so it will be a manual process. This process will be done by the user in order to activate a repo. When a user selects to Activate a specific repo, Woddpecker UI will open a pop up that will report the webhook URL that the user will have to register at Radicle’s side. The repo will be consider as active at this moment. Te same approach will be followed at the deactivation case.
6. git commands will be used for managing repos.

So, do you think that there will be any concern for accepting a forge's implementation that follows this approach?

**Reply by qwerty287** [reply]
Yes, I have some issues with this:

1. OAuth support
2. Webhook enabling via API

I don't like having workarounds for them. Both are pretty essential functions (especially missing oauth, this is a reason why we removed Gogs support).

Some time ago, I had an idea for additions like this: Support Go plugins for some features. If we implement this, it could be possible to register own forges without needing to implement them in woodpecker's core (I'll open a discussion for this in a minute -> #2520).

**Reply by anbraten** [reply]
Adding an alternative auth system in addition to oauth just for radicle scares me a bit as well 🤔

**Reply by mzampetakis** [reply]
OK, thank you for letting us know that both these issues are blockers. We will discuss about adding both as separate plugins in the codebase and report back here.

**Reply by gsaslis** [reply]
Hi folks, 

> I don't like having workarounds for them. Both are pretty essential functions (especially missing oauth, this is a reason why we removed Gogs support).

I do agree here they _are_ both standard functions - for traditional forges. 

The thing is that radicle is **fundamentally** different (pun skipped) from other forges,  because of its p2p nature. In radicle peers (i.e. users) - and not service providers, as is the case for other forges - are expected to deploy the node (and accompanying HTTP API). 

This means that in many cases the node (and HTTP API) are **not publicly accessible on the internet**. In fact, the scenario we had in mind for integration - that @mzampetakis laid out above - involved deploying a radicle node and API together with woodpecker (e.g. in same VM or same local network, etc. etc.) 

What's more, it is not always the case that a web interface (to which the user would be redirected to/from) is even deployed! It is entirely possible to use radicle through the CLI, or an IDE plugin.  

I am far from an expert in auth* flows, so please do correct me if I'm missing something obvious here (I might very well be! 😅), but, with this context in mind, I am wondering: does it still make sense to set OAuth as a requirement? 

As for the webhook, I once again 100% agree that expecting an API endpoint to set some value in a database is also standard method. However, we don't have a database. :| All of the radicle data model lives in git itself. For webhooks in particular, we were thinking of simply storing them in `.radicle/webhooks.yml`, inside the git repo that webhooks are being enabled for, and encrypting the contents of that file with the public keys of the project _delegates_ (i.e. maintainers), so that this information is only readable by them. Having an API endpoint that creates a commit is certainly possible, but it feels a little less standard - which is why I'd still prefer if users could just commit the file to git directly and enable the integration manually.  

To sum up: 
I do understand we are asking for exceptions here and I am generally not in favour of these either, so it is perfectly fine and understandable if woodpecker maintainers prefer not to allow these for this integration. I just thought I'd ask anyway, considering how ... radically different the situation is here (sorry, couldn't resist in the end 😇 ). 

In any case, thanks for your time and attention!

**Reply by qwerty287** [reply]
Hi!

Let me comment on some points you mentioned.
1. OAuth: Actually, OAuth can be implemented without web ui, everything you need is a http service that's accessible from woodpecker. I.e. it works if it's not publicly accessible but accessible from woodpecker. Also, you don't need a web ui, you might not be able to let users control which oauth apps are allowed to access their account, but in general this is possible.
2. Webhooks: While manual enabling is totally possible (or even using woodpecker without any webhooks), I'm still not sure of I'd add a forge which requires this directly to the main repo. Webhooks are the most common and for most forges easy way to automatically run pipelines, and adding a lot of code just to support a special forge is not the best idea I think, forges should be as unspecific in the code as possible, except for their own specific implementation in it's own package (I.e. they implement the interface, but that's almost the full forge-specific code).

I still think that one of the best solutions is to support some kind of plugins (like Go plugins, see #2520) that allows other users to write the forges themselves without having to integrate it in woodpecker's main repo, which is especially helpful for forges like Radical that are missing features and are therefore not accepted to be added to the main repo.

**Reply by 6543** [reply]
Well I would say we should not make the two points a hard block.

For Oauth ... you could create some kind of mock in the forge driver and return a token set beforhand.

In the long run I would recomend to have it properly.

For the webhook enabling, thats not a great blocker ... if your forge target advanced devs ... the process can also be a bit more complicated.

In both cases I would argue radicle with a webui could imitate the behaviours needed for woodpecker and do it completly different

**Reply by qwerty287** [reply]
I agree in general, but I'm kind of against adding this the main repo (but rather build a plugin system and implement it this way) for the following reasons:
1. It probably does not have many users -> more code with not many advantages. If we do this for every forge, we'll have hundreds of them in the main binary which makes maintaining harder and increases binary sizes.
2. Yes, it is _possible_ to not use webhooks and oauth but I think we should have some principles and requirements for forges and these two are the most important points I think.

**Reply by gsaslis** [reply]
thanks both! Helpful feedback here! 

I do agree the plugin approach fits quite nicely here. Is there already an example we can follow?

**Reply by qwerty287** [reply]
Not yet, woodpecker does not even support it yet, it's just an approach. It's on my todo list after we released 2.0 though :)

**Reply by qwerty287** [reply]
I actually just had nothing else to do, so I started working on in and I got a working prototype already, so I'll see when I can finish it and open a PR.

**Reply by mzampetakis** [reply]
Thank you all for your feedback. It's really valuable! 🙌🏻 
@gsaslis made a pretty good abstraction on how radicle works and how we have designed its integration with woodpecker.

I do understand the bottlenecks about oAuth and webhook enabling process, as both of them are part of the backbone of woodpecker's forges integrations. 

Supporting the Radicle forge using Plugins seems an appealing approach and it might be a good fit for this case. My only concern is how - but most importantly when - woodpecker plans to support them. Thankfully, @qwerty287 started working on this 🎉 , so, please let me know if I can provide any kind of assistance on that - I would love to! 

As for the implemented plugins (such as this one we are curenlty discussing or any other that might appear in the near future): are there any plans for them to be integrated officially with woodpecker. I mean, could they be offered through a centralised point as woodpecker does [with plugins](https://woodpecker-ci.org/plugins) right now so user's could be informed officially by their existence and support? This is definitely something that could come later on but having the maintainers' feedback is pretty valuable.

> Maybe these types of _plugins_ have to come with a different naming comparing to the existing ones that are used for enriching pipelines' tasks. 🤔

**Reply by qwerty287** [reply]
> Maybe these types of plugins have to come with a different naming comparing to the existing ones that are used for enriching pipelines' tasks. 🤔

Yeah I called them "addons" in my prototype.


And yes, when we implement/add this, I'd create an index similar to the plugin index to find these addons.

About the time plan: Probably it will be available in the next minor release 2.1.0, but I can't give you an exact time - I need to finish it first and open a PR then.

**Reply by mzampetakis** [reply]
These are great! I did not expect an exact time by any means 😉 
Thanks!

**Comment by 6543** [comment]
I propose that the ideas are almost ready ...

So why not draft a pull and figure out the painpoints, so we can address them on both sides while they apeare?

**Reply by mzampetakis** [reply]
For the records here is the PR which is now merged: https://github.com/woodpecker-ci/woodpecker/pull/2751

**Comment by mzampetakis** [comment]
I have faced an issue where the type of the pull request RemoteForgeID is strictly expected to be an integer (int64) instead of a more generic string type that it is used for other RemoteForgeIDs. At radicle's side a Pull request ID (called patch) it is of string type.
I have describe this case in another issue: https://github.com/woodpecker-ci/woodpecker/discussions/2746 but it's related to this implementation. 

I would like to have your feedback  on this as it seems that it does not follow the other RemoteForgeIDs' types which results to a non-uniform approach for this type.

**Reply by mzampetakis** [reply]
Issue addressed by https://github.com/woodpecker-ci/woodpecker/pull/2823

**Comment by mzampetakis** [comment]
Now that there is the addon support we can move forward and implement the Radicle forge. 🎉
Thanks for getting this support done! ❤️

Is there any designated repo for adding such implementations, or there will be some official listing for 3rd party implementation?

**Reply by 6543** [reply]
Not jet. I still think we should pull the code into the main repo if everything is working.

Or at least add a dedicated entry how to setup in the official docs next to gitea, github & gitlab ...

**Reply by mzampetakis** [reply]
Maybe it could be added as normal forge at `server/forge` and properly documented as addon under https://woodpecker-ci.org/docs/administration/forges/overview.

Will this work for other backend addons? Is there any relevant place for them?

> I am not sure if this is the right place to discuss this. If so, please feel free to propose another place for discussing this topic.

**Reply by qwerty287** [reply]
> Not jet. I still think we should pull the code into the main repo if everything is working.

The only reason for adding addons is that they're not in the main repo, so I wouldn't add it.

We should create an index for addons, but that's not existing yet.

Proper docs can then be done similar to plugins with a MD file that is pulled into our website but maintained by add-on authors.

> Maybe it could be added as normal forge at server/forge and properly documented as addon under https://woodpecker-ci.org/docs/administration/forges/overview.

What do you mean? If adding it to the main repo it *must* be documented there otherwise in a separate add-on registry.

This discussion however, shouldn't stop you from starting the implementation: Switching between addon and adding to main repo is just about changing a few setup code lines, but everything else should stay the same.

**Reply by mzampetakis** [reply]
I have already started the implementation, but still haven't migrated to an addon type (I am integrating it directly to woodpecker repo).

> What do you mean? If adding it to the main repo it must be documented there otherwise in a separate add-on registry.

This is what I mean! The other solution is to be implemented and maintained by add-on authors and listed in some central index as there is one for plugins.

**Reply by qwerty287** [reply]
> The other solution is to be implemented and maintained by add-on authors and listed in some central index as there is one for plugins.

That's the plan from my side. It just needs to be done ;)

**Reply by mzampetakis** [reply]
Please let me know if there is any repo that I could contribute to, in order to assist on the creation of the aforementioned central index page!

**Reply by 6543** [reply]
I would start with adding a [cookbook](https://github.com/woodpecker-ci/woodpecker/tree/main/docs/cookbook) ... how to do radice + woodpecker.

that's a more block like aproach and as it is still fresh new forge plugin this should fit best.

if it did run & is deployed without issues for some time ... we can add it to the forges overview.

as for go-plugins that can come seperated ... for now we could just use the awesome page.

we just have to make sure go-plugins are not plugins ... and if not pointed out clearly it will be a mess for users.
so I would propose to call go-plugins addons or so

**Reply by qwerty287** [reply]
> as for go-plugins that can come seperated ... for now we could just use the awesome page.

I also think the awesome page is fine for now, later I'd add a separate index similar to plugins.

> so I would propose to call go-plugins addons or so

That's how they called already :)

**Comment by mzampetakis** [comment]
We can close this discussion as woodpecker now supports [addons](https://woodpecker-ci.org/docs/administration/addons/overview) that allows us to develop an external forge addon for radicle.

Thank you all!
